Client/game_sa/CVehicleSA.cpp

```diff
@@ -1501,27 +1502,118 @@ void CVehicleSA::SetGravity(const CVector* pvecGravity)
+        {
+            auto* bikeInterface = static_cast<CBikeSAInterface*>(GetVehicleInterface());
+            if (!bikeInterface)
+                return false;
+
+            if (nodeIndex >= sizeof(bikeInterface->m_apModelNodes) / sizeof(RwFrame*))
+                return false;
+
+            nodesOffset = OFFSET_CBike_Nodes;
+            if (nodeIndex != 1)
+                break;
+
+            // Set the correct "bike_chassis" frame for bikes
+            defaultBikeChassisFrame = bikeInterface->m_apModelNodes[1];
+            if (defaultBikeChassisFrame && std::strcmp(defaultBikeChassisFrame->szName, "chassis_dummy") == 0)
+            {
+                RwFrame* correctChassisFrame = ((RwFrame * (__cdecl*)(RpClump*, const char*))FUNC_CClumpModelInfo_GetFrameFromName)(bikeInterface->m_pRwObject, "chassis");
+                if (correctChassisFrame)
+                    bikeInterface->m_apModelNodes[1] = correctChassisFrame;
+            }
+            break;
+        }
+        case VehicleClass::BOAT:
+        {
+            auto* boatInterface = static_cast<CBoatSAInterface*>(GetVehicleInterface());
+            if (!boatInterface)
+                return false;
+
+            if (nodeIndex >= sizeof(boatInterface->pBoatParts) / sizeof(RwFrame*))
+                return false;
+
+            nodesOffset = OFFSET_CBoat_Nodes;
+            break;
+        }
+        default:
+            return false;
+    }
+
+    // Patch nodes array in CAutomobile::SpawnFlyingComponent
+    MemPut(0x6A85B3, nodesOffset);
+    MemPut(0x6A8631, nodesOffset);
+
+    DWORD dwInterface = (DWORD)GetInterface();
     DWORD dwFunc = FUNC_CAutomobile__SpawnFlyingComponent;
+    DWORD dwReturn;
     _asm
     {
-        mov     ecx, dwThis
-        push    ui_2
-        push    i_1
-        call    dwFunc
-        mov     dwReturn, eax
+        mov ecx, dwInterface
+        push collisionType
+        push nodeIndex
+        call dwFunc
+        mov dwReturn, eax
     }
 
-    CObject* pObject = NULL;
-    if (dwReturn)
+    // Restore default nodes array in CAutomobile::SpawnFlyingComponent
+    MemPut(0x6A85B3, 0x648);
+    MemPut(0x6A8631, 0x648);
+
+    // Restore default chassis frame for bikes
+    if (nodeIndex == 1 && defaultBikeChassisFrame)
     {
-        SClientEntity<CObjectSA>* pObjectClientEntity = pGame->GetPools()->GetObject((DWORD*)dwReturn);
-        pObject = pObjectClientEntity ? pObjectClientEntity->pEntity : nullptr;
+        auto* bikeInterface = static_cast<CBikeSAInterface*>(GetVehicleInterface());
+        if (bikeInterface && bikeInterface->m_apModelNodes)
+            bikeInterface->m_apModelNodes[1] = defaultBikeChassisFrame;
     }
-    return pObject;
+
+    if (removalTime <= -1)
+        return true;
+
+    auto* componentObject = reinterpret_cast<CObjectSAInterface*>((DWORD*)dwReturn);
+    if (!componentObject)
+        return;
+
+    int CTimer_ms = *(int*)0xB7CB84; // CTimer::m_snTimeInMilliseconds
```

Comment:
From what I can see it should be DWORD (or uint32_t). See gta_reversed and CGameSA line 192.

Reply:
Yeah, that's right